pkg/kv/kvserver/client_manual_proposal_test.go

```diff
@@ -0,0 +1,272 @@
+			defer tc.Stopper().Stop(ctx)
+			require.NoError(t, tc.ServerConn(0).QueryRow(`SELECT count(1) FROM system.settings`).Err())
+			t.Log("read system.settings")
+		}()
+
+	}
+
+	st := cluster.MakeTestingClusterSettings()
+	eng, err := storage.Open(ctx, storage.Filesystem(p), st)
+	require.NoError(t, err)
+	defer eng.Close()
+
+	// Determine LastIndex, LastTerm, and next MaxLeaseIndex by scanning
+	// existing log.
+	it := raftlog.NewIterator(rangeID, eng, raftlog.IterOptions{})
+	defer it.Close()
+	rsl := logstore.NewStateLoader(rangeID)
+	lastIndex, err := rsl.LoadLastIndex(ctx, eng)
+	require.NoError(t, err)
+	t.Logf("loaded LastIndex: %d", lastIndex)
+	ok, err := it.SeekGE(lastIndex)
+	require.NoError(t, err)
+	require.True(t, ok)
+
+	var lai kvpb.LeaseAppliedIndex
+	var lastTerm uint64
+	require.NoError(t, raftlog.Visit(eng, rangeID, lastIndex, math.MaxUint64, func(entry raftpb.Entry) error {
+		ent, err := raftlog.NewEntry(it.Entry())
+		require.NoError(t, err)
+		if lai < ent.Cmd.MaxLeaseIndex {
+			lai = ent.Cmd.MaxLeaseIndex
+		}
+		lastTerm = ent.Term
+		return nil
+	}))
+
+	sl := stateloader.Make(rangeID)
+	lease, err := sl.LoadLease(ctx, eng)
+	require.NoError(t, err)
+
+	for batchIdx := 0; batchIdx < batches; batchIdx++ {
+		t.Logf("batch %d", batchIdx+1)
+		b := storage.NewOpLoggerBatch(eng.NewBatch())
+		defer b.Batch.Close()
+
+		var ents []raftpb.Entry
+		for i := 0; i < entsPerBatch; i++ {
+			lai++
+			c := &kvpb.ProbeRequest{}
+			resp := &kvpb.ProbeResponse{}
+			c.Key = keys.LocalMax
+
+			cmd, ok := batcheval.LookupCommand(c.Method())
+			require.True(t, ok)
+			// NB: this should really operate on a BatchRequest. We need to librarize
+			// evaluateBatch:
+			// https://github.com/cockroachdb/cockroach/blob/9c09473ec9da9d458869abb3fe08a9db251c9291/pkg/kv/kvserver/replica_evaluate.go#L141-L153
+			// The good news is, this is already in good shape! Just needs to be moved
+			// to a leaf package, like `batcheval`.
+			// To use the "true" logic we want this for sure, and probably even the
+			// caller of evaluateBatch and a few more levels. Pulling it out until
+			// there's nothing left basically.
+
+			evalCtx := &batcheval.MockEvalCtx{}
+			args := batcheval.CommandArgs{
+				EvalCtx:     evalCtx.EvalContext(),
+				Header:      kvpb.Header{},
+				Args:        c,
+				Stats:       &enginepb.MVCCStats{},
+				Uncertainty: uncertainty.Interval{},
+			}
+			res, err := cmd.EvalRW(ctx, b, args, resp)
+			require.NoError(t, err)
+			// NB: there's more stuff in evaluateProposal that would need to be lifted
+			// here:
+			// https://github.com/cockroachdb/cockroach/blob/f048ab082c58ec0357b2ecad763606ef64faa3b7/pkg/kv/kvserver/replica_proposal.go#L842-L869
+			res.WriteBatch = &kvserverpb.WriteBatch{Data: b.Repr()}
+			res.LogicalOpLog = &kvserverpb.LogicalOpLog{Ops: b.LogicalOps()}
+
+			// End of evaluation. Start of "proposing".
+
+			// TODO: the "requires consensus" logic is not reusable, make it so:
```

Comment:
Is there an issue? More generally, is this test going to be unskipped at some point in some other form? Do we need to avoid this test being dead code?

If this test is intended as a demo, link it in the reverse direction from the issue that it serves? So that it's not forgotten.

Reply:
Filed https://github.com/cockroachdb/cockroach/issues/105177 and cross-referring from the comments to that issue and back.